### ISSUE TITLE
`OrderedDict` vararg constructors don't make use of type promotion

### ISSUE THREAD

**Issue by omus** [JuliaCollections/OrderedCollections.jl#47]
When using the `OrderedDict` vararg constructors with pairs using different types the type `Any` will be used instead of using type promotion. This can result in failures when interacting with other functions expecting a narrower type:

```julia
julia> using Intervals, TimeZones, OrderedCollections, DataFrames

julia> args = (
               :interval => Interval{ZonedDateTime},  # UnionAll
               :date => ZonedDateTime,  # DataType
       );

julia> eltype(OrderedDict(args...))
Pair{Symbol,Any}

julia> eltype([args...])
Pair{Symbol,Type}

julia> schema = OrderedDict(args...);

julia> DataFrame(collect(values(schema)), collect(keys(schema)), 0)
ERROR: MethodError: no method matching DataFrame(::Array{Any,1}, ::Array{Symbol,1}, ::Int64)
Closest candidates are:
  DataFrame(::AbstractArray{T,1}, ::AbstractArray{Symbol,1}, ::Integer; makeunique) where T<:Type at /Users/omus/.julia/packages/DataFrames/S3ZFo/src/dataframe/dataframe.jl:210
  DataFrame(::AbstractArray{T,1} where T, ::AbstractArray{Symbol,1}; makeunique, copycols) at /Users/omus/.julia/packages/DataFrames/S3ZFo/src/dataframe/dataframe.jl:178
  DataFrame(::Union{Array{Any,1}, Array{AbstractArray{T,1} where T,1}}, ::DataFrames.Index; copycols) at /Users/omus/.julia/packages/DataFrames/S3ZFo/src/dataframe/dataframe.jl:96
  ...
Stacktrace:
 [1] top-level scope at REPL[16]:1

 julia> schema = OrderedDict{Symbol, Type}(args...);

julia> DataFrame(collect(values(schema)), collect(keys(schema)), 0)
0×2 DataFrame
```

**Comment by oxinabox** [comment]
It does indeed seem that our behavour is different from that of `Dict`

Fun fact:
Your code breaks `show` in the current version of `DataFrames` if run with a `Dict` instread of an `OrderedDict`.


```
julia> schema = Dict(args...);

julia> DataFrame(collect(values(schema)), collect(keys(schema)), 0)
Error showing value of type DataFrame:
ERROR: TypeError: in typeassert, expected DataType, got Type{Interval{ZonedDateTime,L,R} where R<:Bound where L<:Bound}
Stacktrace:
 [1] compacttype(::Type, ::Int64, ::Bool) at /Users/oxinabox/.julia/packages/DataFrames/e3P4n/src/abstractdataframe/show.jl:100
 [2] compacttype(::Type) at /Users/oxinabox/.julia/packages/DataFrames/e3P4n/src/abstractdataframe/show.jl:69
 [3] getmaxwidths(::DataFrame, ::IOContext{REPL.Terminals.TTYTerminal}, ::UnitRange{Int64}, ::UnitRange{Int64}, ::Symbol, ::Nothing, ::Bool, ::Base.GenericIOBuffer{Array{UInt8,1}}) at /Users/oxinabox/.julia/packages/Data
Frames/e3P4n/src/abstractdataframe/show.jl:182
 [4] _show(::IOContext{REPL.Terminals.TTYTerminal}, ::DataFrame; allrows::Bool, allcols::Bool, splitcols::Bool, rowlabel::Symbol, summary::Bool, eltypes::Bool, rowid::Nothing) at /Users/oxinabox/.julia/packages/DataFrame
s/e3P4n/src/abstractdataframe/show.jl:589
 [5] #show#445 at /Users/oxinabox/.julia/packages/DataFrames/e3P4n/src/abstractdataframe/show.jl:666 [inlined]
 [6] show(::IOContext{REPL.Terminals.TTYTerminal}, ::MIME{Symbol("text/plain")}, ::DataFrame; allrows::Bool, allcols::Bool, splitcols::Bool, rowlabel::Symbol, summary::Bool, eltypes::Bool) at /Users/oxinabox/.julia/packa
ges/DataFrames/e3P4n/src/abstractdataframe/io.jl:56
 [7] show(::IOContext{REPL.Terminals.TTYTerminal}, ::MIME{Symbol("text/plain")}, ::DataFrame) at /Users/oxinabox/.julia/packages/DataFrames/e3P4n/src/abstractdataframe/io.jl:56
 [8] display(::REPL.REPLDisplay, ::MIME{Symbol("text/plain")}, ::Any) at /usr/local/src/julia/julia-1.4/usr/share/julia/stdlib/v1.4/REPL/src/REPL.jl:137
 [9] display(::REPL.REPLDisplay, ::Any) at /usr/local/src/julia/julia-1.4/usr/share/julia/stdlib/v1.4/REPL/src/REPL.jl:141
 [10] display(::Any) at ./multimedia.jl:323
```

**Comment by oxinabox** [comment]
Just checked and this was **not** fixed by #40
